lib/wibox/init.lua

```diff
@@ -61,6 +63,72 @@ function wibox:find_widgets(x, y)
     return self._drawable:find_widgets(x, y)
 end
 
+-- This function is called on every redraw after the background has been drawn,
+-- but before the actual widget was drawn. We use this to draw an anti-aliased
+-- clipping shaped (the edge between the border and the background is
+-- anti-aliased)
+local function draw_hook(self, cr)
+    local shape = self._shape
+    if not shape then return end
+
+    local geo = self:geometry()
+    cr:set_source(gcolor(self.border_color))
+
+    -- We want to invert the shape (draw to anything that is not inside of it).
+    -- To fake this, we use EVEN_ODD and a rectangle containing the full
+    -- surface. Pixels in the border will be inside the big rectangle, so they
+    -- are contained once in the path. Pixels in the "content area" are
+    -- additionally contained in the shape, so twice.
+    cr:set_fill_rule(cairo.FillRule.EVEN_ODD)
+    cr:rectangle(0, 0, geo.width, geo.height)
+    shape(cr, geo.width, geo.height)
+
+    cr:fill()
+end
+
+function wibox:_apply_shape()
+    local shape = self._shape
+
+    if not shape then
+        self.shape_bounding = nil
+        self.shape_clip = nil
+        return
+    end
+
+    local geo = self:geometry()
+    local bw = self.border_width
+
+    -- First handle the bounding shape (things including the border)
+    local img = cairo.ImageSurface(cairo.Format.A1, geo.width + 2*bw, geo.height + 2*bw)
+    local cr = cairo.Context(img)
+
+    shape(cr, geo.width + 2*bw, geo.height + 2*bw)
```

Comment:
Umm, eh, isn't this my first comment here? I just pointed out a little bug in this patch. Note that my Radical module implementation of this has the bug (because I am lazy) and someone complained on IRC about it (a long time ago). It does look weird:

![slider10](https://cloud.githubusercontent.com/assets/340384/19835292/c5052276-9e57-11e6-81b4-07389027e750.png)

(not using your patch). See the `\` part of the `''\/''` I hope you see the issue with doing this. My version of this PR had this corner case covered.